marine/marine.py

```diff
@@ -154,3 +175,37 @@ def _add_or_get_filter(
 
     def __del__(self):
         self._marine.destroy_marine()
+
+    @staticmethod
+    def replace_macros(
+        fields: List[str], macros: Optional[Dict[str, List[str]]]
+    ) -> Tuple[List[str], List[str]]:
+        demacroed_fields = []
+        used_macros = []
+
+        if not macros:
+            return fields, used_macros
+
+        for field in fields:
+            if field in macros.keys():
+                used_macros.append(field)
+                demacroed_fields.extend(macros[field])
+            else:
+                demacroed_fields.append(field)
+
+        return demacroed_fields, used_macros
+
+    @staticmethod
+    def restore_macros(
+        result: Dict[str, str], macros: Optional[Dict[str, List[str]]], used_macros: List[str]
+    ) -> Dict[str, str]:
+        for macro in used_macros:
+            macro_value = ""
+            for value in macros.get(macro, []):
+                pop_value = result.pop(value, "")
+                if pop_value != "" and macro_value == "":
+                    macro_value = pop_value
+
+            result[macro] = macro_value
```

Comment:
Don't mutate your input please.
That's highly confusing and very error prone.
Mutation should be rare and explicit.